p2p/starknet/block_bodies.go

```diff
@@ -0,0 +1,164 @@
+package starknet
+
+import (
+	"slices"
+
+	"github.com/NethermindEth/juno/adapters/core2p2p"
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/p2p/starknet/spec"
+	"github.com/NethermindEth/juno/utils"
+	"google.golang.org/protobuf/proto"
+)
+
+type streamState int
+
+const (
+	_ streamState = iota
+
+	sendDiff // initial
+	sendClasses
+	sendBlockFin
+	terminal // final state
+)
+
+type blockBodyIterator struct {
+	bcReader blockchain.Reader
```

Comment:
I don't think we need to pass the blockchain Reader here as we only use it for 2 calls. Usually, it is not a good idea to pass alarge interface when it is not fully utilised (we should break up the Reader into smaller interfaces, this is for another pr, ofcourse).

Passing `core.StateUpdate` and `core.StateReader` should be enough. We only use the blockchain reader for a single block and the iterator is also created per block, so I don't think it is wise to pass a stuct that is intended to provide data for multiple blocks.